core/drivers/pl022_spi.c

```diff
@@ -0,0 +1,532 @@
+	*num_rxpkts = j;
+}
+
+static void pl022_tx8(struct spi_chip *chip, uint8_t *wdat,
+	uint32_t num_txpkts)
+{
+	uint32_t i;
+	struct pl022_data *pd = container_of(chip, struct pl022_data, chip);
+
+	gpio_set_value(pd->cs_gpio_pin, GPIO_LEVEL_LOW);
+
+	for (i = 0; i < num_txpkts; i++) {
+		if (read32(pd->base + SSPSR) & SSPSR_TNF) {
+			/* tx 1 packet */
+			write8(wdat[i], pd->base + SSPDR);
+		} else {
+			DMSG("TX FIFO full.. waiting..\n");
+			while ((read32(pd->base + SSPSR) & SSPSR_TNF) == 0)
+				tee_time_wait(1);
+
+			DMSG("TX FIFO available.. resuming..\n");
+			i--; /* retry current packet in next loop */
+		}
+	}
+
+	gpio_set_value(pd->cs_gpio_pin, GPIO_LEVEL_HIGH);
+}
+
+static void pl022_tx16(struct spi_chip *chip, uint16_t *wdat,
+	uint32_t num_txpkts)
+{
+	uint32_t i;
+	struct pl022_data *pd = container_of(chip, struct pl022_data, chip);
+
+	gpio_set_value(pd->cs_gpio_pin, GPIO_LEVEL_LOW);
+
+	for (i = 0; i < num_txpkts; i++) {
+		if (read32(pd->base + SSPSR) & SSPSR_TNF) {
+			/* tx 1 packet */
+			write16(wdat[i], pd->base + SSPDR);
+		} else {
+			DMSG("TX FIFO full.. waiting..\n");
+			while ((read32(pd->base + SSPSR) & SSPSR_TNF) == 0)
+				tee_time_wait(1);
+
+			DMSG("TX FIFO available.. resuming..\n");
+			i--; /* retry current packet in next loop */
+		}
+	}
+
+	gpio_set_value(pd->cs_gpio_pin, GPIO_LEVEL_HIGH);
+}
+
+static void pl022_rx8(struct spi_chip *chip, uint8_t *rdat,
+	uint32_t *num_rxpkts)
+{
+	uint32_t j = 0;
+	struct pl022_data *pd = container_of(chip, struct pl022_data, chip);
+
+	gpio_set_value(pd->cs_gpio_pin, GPIO_LEVEL_LOW);
+
+	do {
+		while (read32(pd->base + SSPSR) & SSPSR_RNE)
+			rdat[j++] = read8(pd->base + SSPDR);
+
+	} while (read32(pd->base + SSPSR) & SSPSR_BSY);
+
+	gpio_set_value(pd->cs_gpio_pin, GPIO_LEVEL_HIGH);
+	*num_rxpkts = j;
+}
+
+static void pl022_rx16(struct spi_chip *chip, uint16_t *rdat,
+	uint32_t *num_rxpkts)
+{
+	uint32_t j = 0;
+	struct pl022_data *pd = container_of(chip, struct pl022_data, chip);
+
+	gpio_set_value(pd->cs_gpio_pin, GPIO_LEVEL_LOW);
+
+	do {
+		while (read32(pd->base + SSPSR) & SSPSR_RNE)
+			rdat[j++] = read16(pd->base + SSPDR);
+
+	} while (read32(pd->base + SSPSR) & SSPSR_BSY);
+
+	gpio_set_value(pd->cs_gpio_pin, GPIO_LEVEL_HIGH);
+	*num_rxpkts = j;
+}
+
+static void pl022_print_peri_id(struct pl022_data *pd __maybe_unused)
+{
+	DMSG("Expected: 0x 22 10 #4 0");
+	DMSG("Read: 0x %x %x %x %x\n",
+		read32(pd->base + SSPPeriphID0),
+		read32(pd->base + SSPPeriphID1),
+		read32(pd->base + SSPPeriphID2),
+		read32(pd->base + SSPPeriphID3));
+}
+
+static void pl022_print_cell_id(struct pl022_data *pd __maybe_unused)
+{
+	DMSG("Expected: 0x 0D F0 05 B1");
+	DMSG("Read: 0x %x %x %x %x\n",
+		read32(pd->base + SSPPCellID0),
+		read32(pd->base + SSPPCellID1),
+		read32(pd->base + SSPPCellID2),
+		read32(pd->base + SSPPCellID3));
+}
+
+static void pl022_sanity_check(struct pl022_data *pd)
+{
+	DMSG("SSPB2BTRANS: Expected: 0x2. Read: 0x%x\n",
```

Comment:
could check here that  fields expected by `pl022_configure()` are initialized: non null, assuming pl022_init() resets the pl022_data structure. i.e `base`,  and `cs_gpio_pin`